### PR TITLE
Fix for when recommendations are short

### PR DESCRIPTION
**Pull request by saaalik**
## Description
- Fixed `get_rated_bookids` function
- When combining the two recommendation lists, if that list<`recommendations`(9) then my code snippet is executed which fills in remaining books to be recommended, half from cv_bookids and the other half with top rated books of the most common genre among `already_rated_books` and the books yet recommended.
- To recommend most rated genre book, A pickle file was created to store the value of the dictionary storing the same, files used can be accessed [here](https://easyupload.io/m/m87iol)
- Issue #38 

## Affected Dependencies
Changes are made only into `mainapp/views.py` and `helper.py` affecting the personalized-recommendations section of the project

## How has this been tested?
Tested on python shell manually as well as on the main project webpage. Rated 5 different books, when the `best_bookids` was calculated manually, only 5 books were recommended even after summing up bookids from `cv_bookids` and `embedding_bookids` and thus my code snippet executed and added 2 bookids from `cv_bookids` and 2 similar genre books

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [x] My changes do not edit/add any unrequired files.
- [x] My changes are covered by tests.

### CODE REVIEW

File: mainapp/helpers.py [hunk 1]
```diff
@@ -19,6 +22,9 @@
 book_embed_path = os.path.join(settings.STATICFILES_DIRS[0] + '/mainapp/model_files/surprise/book_embedding.npy')
 sim_books_path = os.path.join(settings.STATICFILES_DIRS[0] + '/mainapp/model_files/surprise/sim_books.pickle')
 
+# For Similar-Genre books
+sim_genre_books_path = os.path.join(settings.STATICFILES_DIRS[0] + '/mainapp/model_files/genre_dictionary.pickle')
```
Comment (Praful932) [inline mainapp/helpers.py:26]: Could you explain, Why is this needed? 
The remaining `n` 's genre as discussed should be taken from the user's most common(rated & recommended) genre and return n top books of that genre not present in `recommended & rated`, I think `df_book` already has all what you need

File: mainapp/views.py [hunk 2]
```diff
@@ -58,7 +58,7 @@ def book_recommendations(request):
     random.shuffle(user_ratings)
     best_user_ratings = sorted(user_ratings, key=operator.attrgetter('bookrating'), reverse=True)
 
-    if len(best_user_ratings) < 4:
+    if len(best_user_ratings) < 5:
```
Comment (Praful932) [inline mainapp/views.py:61]: You need to change this back, even if user has `5` ratings it will not pass

**Review by Praful932 — CHANGES_REQUESTED** [review]

File: mainapp/helpers.py (hunk 1 above)
Comment (saaalik) [reply mainapp/helpers.py:26]: The file `genre_dictionary.pickle` contains inverted index of all different genres, sorted in descending order of votes from column `count` in `full_genre.csv`, making the process less time consuming, else the script would have to loop through a lot of books checking for the highest rating and the corresponding genre. If `count` column (genre votes) is not a directly proportional variable to the book_ratings then I can make an inverted index using books.csv such that
```python
genre_index = {genre: [ [rating, book_id], [rating2, book_id2] ] ,
                         genre2: [ [rating, book_id],[rating, book_id],[rating, book_id] ],
                         genre3: [].......}
```
sorted in descending order of rating. Then directly importing it to the function this variable is being used in.

File: mainapp/views.py (hunk 2 above)
Comment (saaalik) [reply mainapp/views.py:61]: Okey, I will revert back this change

File: mainapp/helpers.py (hunk 1 above)
Comment (Praful932) [reply mainapp/helpers.py:26]: Better reuse [this](https://github.com/Praful932/Kitabe/blob/master/mainapp/helpers.py#L98)function, add a parameter `n_books`(default = 16) - For genre page.
Use this function to get total `2*needed_books` (to avoid duplicates)  for any particular genre

**Comment by Praful932** [comment]
@saaalik Re-request a review when the comments are resolved, Thanks

File: mainapp/helpers.py [hunk 3]
```diff
@@ -95,12 +96,12 @@ def get_bookid(raw_id_list):
...
 
 
-def genre_wise(genre, percentile=0.85):
+def genre_wise(genre, n_books=16, percentile=0.85):
     '''
         Returns top genre books according to a cutoff percentile to be listed in Top Books
     '''
-    n_books = 16
     min_genre_book_count = 48
+    cols = ['average_rating', 'book_id']
```
Comment (Praful932) [inline mainapp/helpers.py:104]: This will break `genre` page

File: mainapp/helpers.py [hunk 4]
```diff
@@ -112,7 +113,7 @@ def genre_wise(genre, percentile=0.85):
     qualified = qualified.assign(weighted_rating=W)
     qualified.sort_values('weighted_rating', ascending=False, inplace=True)
 
-    return qualified[cols].head(min_genre_book_count).sample(n_books)
+    return qualified[cols].head(min_genre_book_count).sample(int(2*n_books))
```
Comment (Praful932) [inline mainapp/helpers.py:116]: why `int`, let's pass an already converted `integer` value instead of leaving it to the function

File: mainapp/helpers.py [hunk 5]
```diff
@@ -186,10 +187,34 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
 
     if len(best_bookids) < recommendations:
         # If not enough recommendations
-        best_bookids = best_bookids + cv_bookids[3:(recommendations + 3 - len(best_bookids))]
+        two_n = (recommendations - len(best_bookids))
+        n1, n2 = math.ceil(two_n/2), math.floor(two_n/2)
+        best_bookids_remain = cv_bookids[3: 6+n1]
```
Comment (Praful932) [inline mainapp/helpers.py:192]: If you put `6` because of duplicates, Specify `3 * 2`, for more clarity, +1 if you add comments wherever possible, since for a new contributor it's a bit difficult to catch up

File: mainapp/helpers.py [hunk 6]
```diff
@@ -186,10 +187,34 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
...
+        best_bookids = best_bookids + best_bookids_remain + genre_recomm_bookids
     return best_bookids
 
 
+def most_common_genre_recommendations(best_bookids, already_rated, best_bookids_remain, n):
+    '''
+        Returns n top rated of the most_common_genre among all lists taken as input
+    '''
+    # Final list of bookids to be recommended
+    books = set(best_bookids+list(already_rated))
```
Comment (Praful932) [inline mainapp/helpers.py:204]: You need to find common genre among -  `best_bookids`,`already_rated` & the first half which will be recommended

**Review by Praful932 — CHANGES_REQUESTED** [review]
I have added some comments, resolve them. Also it would be great if you replace wherever `cv` method or variable is mentioned to `tfidf` since we adopted the tfidf approach now.
And rename `best_bookids_remain` to `best_bookids_tfidf` since they are the first half of the remaining `2n` to be recommended, revert any line changes such as `32` in `helpers.py` and remove the pickle file, since we aren't using that now.

File: mainapp/helpers.py [hunk 7]
```diff
@@ -95,11 +96,12 @@ def get_bookid(raw_id_list):
...
 
 
-def genre_wise(genre, percentile=0.85):
+def genre_wise(genre, n_books=16, percentile=0.85):
     '''
         Returns top genre books according to a cutoff percentile to be listed in Top Books
     '''
-    n_books = 16
+    # To return double the number of books required
+    n_books = n_books * 2
```
Comment (Praful932) [inline mainapp/helpers.py:104]: 😅, What I meant is keep this as it is, whenver you pass, pass it `x2`

File: mainapp/helpers.py [hunk 8]
```diff
@@ -170,26 +172,63 @@ def get_book_dict(bookid_list):
...
+    books = set(best_bookids+list(already_rated)+best_bookids_tfidf)
+
+    # Accumulation of all the genres listed in `books` variable
+    genre_frequency = []
+    for book in books:
+        genre_frequency.extend(df_book[df_book['book_id'] == book]['genre'].values[0].split(", "))
+
+    # The most common genre among the bookids in `books` variable
+    most_common_genre = Counter(genre_frequency).most_common(1)[0][0]
+
```
Comment (Praful932) [inline mainapp/helpers.py:220]: @saaalik I could follow till here, you got the most common genre, could you elaborate on the code below and why is there a loop needed?

**Review by Praful932 — CHANGES_REQUESTED** [review]
Delete the extra `pickle` file

File: mainapp/helpers.py (hunk 8 above)
Comment (saaalik) [reply mainapp/helpers.py:220]: genre_wise function will always return 2*n number of books, but if a case came up where the unique sets of bookids are less than n after removing the common bookids from `books` i.e. after 
```python
genre_recommendations = genre_recommendations.difference(books)    #Might return a list of length<n in rare test cases#
```, then the genre_wise function will be executed again, and the weight calculation method always randomizes the output, so the loop will execute till the returned value of genre_wise function is less than n

Comment (Praful932) [reply mainapp/helpers.py:220]: `books`  will always have a unique set since we took care of that, so whatever remaining `n` books we take from whatever genre, just make it `2*n` and take a `difference` and take first `n` books, don't see why loop would be needed

Comment (saaalik) [reply mainapp/helpers.py:218]: Removed the while loop, please review

File: mainapp/helpers.py [hunk 9]
```diff
@@ -112,7 +112,7 @@ def genre_wise(genre, percentile=0.85):
     qualified = qualified.assign(weighted_rating=W)
     qualified.sort_values('weighted_rating', ascending=False, inplace=True)
 
-    return qualified[cols].head(min_genre_book_count).sample(n_books)
+    return qualified[cols].head(min_genre_book_count).sample(2*n_books)
```
Comment (Praful932) [inline mainapp/helpers.py:115]: If you modify this, then `2*n` goes to genre too, pass `2*n` from where you need itself

File: mainapp/helpers.py [hunk 10]
```diff
@@ -170,26 +170,61 @@ def get_book_dict(bookid_list):
...
+    genre_frequency = []
+    for book in books:
+        genre_frequency.extend(df_book[df_book['book_id'] == book]['genre'].values[0].split(", "))
+
+    # The most common genre among the bookids in `books` variable
+    most_common_genre = Counter(genre_frequency).most_common(1)[0][0]
+
+    # Recommendations list, listing n or more unique bookids
+    genre_recommendations = set()
+    genre_recommendations = set(genre_wise(most_common_genre, n).book_id.to_numpy())
```
Comment (Praful932) [inline mainapp/helpers.py:221]: `2*n` here

**Review by Praful932 — CHANGES_REQUESTED** [review]
Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the  `count_vectorizer_recommendations` function name to `tfidf_recommendations`

File: mainapp/helpers.py (hunk 10 above)
Comment (saaalik) [reply mainapp/helpers.py:221]: ooh alright, I'll do that

**Comment by saaalik** [comment]
> Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the `count_vectorizer_recommendations` function name to `tfidf_recommendations`

I don't know how what books will be genreated by genre_wise function for any test cases, because genre_wise function returns different books every time, What I can do is check for the most_common_genre among books i.e.(`set(best_bookids + list(already_rated) + best_bookids_tfidf)`) and cross check the genre of n2 recommended books by the function `most_common_genre_recommendations` like [this...](https://easyupload.io/omoq0k)
In this test case the most common genre is `fiction`

**Comment by Praful932** [comment]
> > Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the `count_vectorizer_recommendations` function name to `tfidf_recommendations`
> 
> I don't know how what books will be genreated by genre_wise function for any test cases, because genre_wise function returns different books every time, What I can do is check for the most_common_genre among books i.e.(`set(best_bookids + list(already_rated) + best_bookids_tfidf)`) and cross check the genre of n2 recommended books by the function `most_common_genre_recommendations` like [this...](https://easyupload.io/omoq0k)
> In this test case the most common genre is `fiction`
Yes exactly you can do that and vary the test cases with edge ones

**Comment by saaalik** [comment]
[These](https://easyupload.io/m/9he5yl) are the test cases for the function, please have a look

**Review by Praful932 — CHANGES_REQUESTED** [review]
Unstage `settings.py`, also post the ss here, it would be easier for us to review, you can directly `ctrl + v` in the PR after taking a snip `ctrl + shift + s` :)

**Comment by Praful932** [comment]
A general guideline for creating TC for your changes
- Assume `already`  has atleast >= 5 books
- `bestbookids` - embeddings can have 0-6 books
- All other variables will change according to the above variables, you already know what values will they be if you fix some values above, so you just have to assert that, similarly make multiple test cases for different scenarios

**Comment by saaalik** [comment]
### TESTCASES for `most_common_genre_recommendations`

Prerequisites
testcase1
testcase2
testcase3
testcase4
testcase5
testcase6
testcase7
testcase8

### TESTCASES for the Counter bit edited to return most common genre based on priority list
[Link](https://easyupload.io/61mj2x) to the py file for testing
Screenshot (259)
Screenshot (260)

**Review by Praful932 — CHANGES_REQUESTED** [review]
I don't see any changes in test files? You need to add the tests in that file and the `most_common_genre_recommendations` seems to be changed? It was fine..

**Comment by saaalik** [comment]
Is testfiles a folder in the repo? I don't understand

**Comment by Praful932** [comment]
@saaalik [This one](https://github.com/Praful932/Kitabe/blob/master/mainapp/tests.py)

File: mainapp/helpers.py (hunk 8 above)
Comment (Praful932) [reply mainapp/helpers.py:221]: Why is the code below changed after you finished with the implementation earlier and were about to write the tests, it was correct then

File: mainapp/helpers.py [hunk 11]
```diff
@@ -29,8 +31,8 @@
 with open(sim_books_path, 'rb') as handle:
     sim_books_dict = pickle.load(handle)
 
-
 cols = ['original_title', 'authors', 'average_rating', 'image_url', 'book_id']
+priority_list = ['fiction', 'fantasy', 'classics', 'contemporary', 'mystery', 'nonfiction', 'paranormal', 'romance', 'history', 'thriller', 'horror', 'memoir', 'comics', 'biography', 'philosophy', 'science', 'crime', 'psychology', 'christian', 'business', 'poetry', 'music', 'religion', 'manga', 'art', 'spirituality', 'cookbooks', 'travel', 'ebooks', 'sports', 'suspense']
```
Comment (Praful932) [inline mainapp/helpers.py:35]: Why is this needed?

File: mainapp/tests.py [hunk 12]
```diff
@@ -209,3 +214,36 @@ def test_user_rated_book_valid(self):
         self.assertEquals(rating.bookrating, valid_bookrating)
         self.assertEquals(rating.user, self.user)
         self.client.logout()
+
+class MostCommonGenreTestCase(TestCase):
```
Comment (Praful932) [inline mainapp/tests.py:218]: We can review this after we are on the same page with the new logic for the recommendations function

File: mainapp/helpers.py (hunk 11 above)
Comment (saaalik) [reply mainapp/helpers.py:35]: okey, so first off we loop through all the book ids that were given as input and append genres to a list.
that list looks something like this 
```python
['art','art','mystery','romance','romance','fiction','fiction']
```
when this list is passed on to Counter(), it returns something like
```python
{'art':2, 'romance':2, 'fiction':2, 'mystery':1}
```
and the Counter(list).most_common(1) returns the very first item from the dictionary
```
[{'art',2)]
```
but because of this in some cases the order of genre changes after Counter(), resulting in test genre and recommended genre being different, failing the test checks.
So after passing the list to Counter()
```python
{'art':2, 'romance':2, 'fiction':2, 'mystery':1}
```
the genre with highest priority, i.e. the greatest book count is recommended. and `priority_list` stores that order from 0 to n in desending order by

1. Slice out the highest value pairs from the dictionary
```python
most_common_dict = {u: v for u, v in genre_count.items() if v == max_value}
# output {'art':2, 'romance':2, 'fiction':2}
```
2. map out the priority list with index values
```python
index_map = {v: i for i, v in enumerate(priority_list)}
# output {'fiction':0, 'fantasy':1, 'classics':2, 'contemporary':3, ........}
```
3. sorting the dictionary pairs on the index value of a genre
```python
final_list = sorted(most_common_dict.items(), key=lambda pair: index_map[pair[0]])
# output [('fiction',2), ('romance',2), ('art',2)]
```
4. return the highest priority genre
```python
most_common_genre = final_list[0][0]
# output 'fiction'
```

Comment (Praful932) [reply mainapp/helpers.py:35]: I don't think it is a good practise to hardcode the priority list, instead what you can do to remove the conflict is order the conflict same count genre according to their alphabetical order(take the first one) or according to the count of that genre in the database(take the max one).

File: mainapp/tests.py (hunk 12 above)
Comment (Praful932) [reply mainapp/tests.py:218]: Regarding the test don't hardcode the book ids, just take the first 5, then the next 5 and so on as we discussed in discord from the dataframe.
Btw do ping us if you need help

**Comment by Praful932** [comment]
Hi @saaalik could you check the tests I had added some comments

File: mainapp/helpers.py (hunk 11 above)
Comment (saaalik) [reply mainapp/helpers.py:35]: I have removed the hard coded list and I am implementing a better way to do the same, will make a pr by EOD

**Comment by saaalik** [comment]
> Hi @saaalik could you check the tests I had added some comments

I have removed the hard coded priority list as well as the bookids instead randomised their selection and also changed the code such that it finds the book count of the highest frequency books `most_common_dict`, and make a dictionary with key as genre and value as book count, after that the genre with highest book count is selected

File: mainapp/helpers.py [hunk 13]
```diff
@@ -170,26 +170,78 @@ def get_book_dict(bookid_list):
...
+
+    if genre_frequency:
+        # The most common genre among the bookids in `books` variable
+        genre_count = dict(Counter(genre_frequency))
+        max_value = max(genre_count.values())
+
+        # Cut out dictionary containing the highest frequency of genre
+        most_common_dict = {u: v for u, v in genre_count.items() if v == max_value}
+
+        # Sort genre with same frequency based on priority list
```
Comment (Praful932) [inline mainapp/helpers.py:224]:
```suggestion
```

File: mainapp/tests.py (hunk 12 above)
Comment (Praful932) [reply mainapp/tests.py:219]: Make a wrapper for all the tests which takes in parameters to avoid duplicate code

**Review by Praful932 — CHANGES_REQUESTED** [review]
Update your fork

**Review by Praful932 — CHANGES_REQUESTED** [review]
Checks are failing :/

**Comment by saaalik** [comment]
I've left the percentile input to default, I can't understand its functionality, what should be the ideal value for percentile in `genre_wise` function?

**Comment by Praful932** [comment]
Some checks are failing, you can run the checks on local, see the contributing.md file.
The percentile is just a cutoff, you can think of it as top books that are liked by 85% people, you can leave it at the default value

File: mainapp/tests.py [hunk 14]
```diff
@@ -211,6 +218,61 @@ def test_user_rated_book_valid(self):
...
+        self.template(10, 6, 2)
+        self.template(10, 6, 3)
+        self.template(10, 6, 4)
+        self.template(10, 7, 1)
+        self.template(10, 7, 2)
+        self.template(10, 7, 3)
+        self.template(10, 8, 1)
+        self.template(10, 8, 2)
+        self.template(10, 9, 1)
+        self.template(10, 10, 0)
```
Comment (Praful932) [inline mainapp/tests.py:259]:
```python
test_cases = [(10, 5, 1), (10, 5, 2), (10, 5, 2), (10, 5, 3), (10, 5, 4),(10, 5, 5), (10, 6, 1), (10, 6, 1), (10, 6, 2), (10, 6, 3), (10, 6, 4),
          (10, 7, 1), (10, 7, 2), (10, 7, 3), (10, 8, 1), (10, 8, 2), (10, 9, 1), (10, 10, 0)]
for tnum, already_slice, bestbookids_slice in test_cases:
    self.template(tnum, already_slice, bestbookids_slice )
```
Also add a docstring to template, so in future, everyone is aware what the params mean

**Review by Praful932 — CHANGES_REQUESTED** [review]
Added some comments

**Comment by saaalik** [comment]
Added the changes, please review

**Review by Praful932 — CHANGES_REQUESTED** [review]
Changes look fine to me, but the tests are failing :/
Also could you attach the full output of `python manage.py test`

**Comment by Praful932** [comment]
Hi @saaalik could you update the fork, we'll make the changes and merge it

**Comment by saaalik** [comment]
I have updated the fork, I am sorry I wasn't able to fix the issue properly, this was my first time contributing to a project of this scale. But I couldn't find what I was doing wrong to get an error in some test cases

**Comment by Praful932** [comment]
No probs

**Review by maneprajakta — APPROVED** [review]
LGTM :rocket:

**Review by Praful932 — APPROVED** [review]
🚀
